### ISSUE TITLE
wrong diagnostics when use `cipher::KeyInit` and `cipher::KeyIvInit` together

### ISSUE THREAD

**Issue by avalon1610** [rust-lang/rust-analyzer#18176]
**rust-analyzer version**: 0.3.2121-standalone

**rustc version**: rustc 1.81.0 (eeb90cda1 2024-09-04)

**editor or extension**: VSCode 1.92.1, Extension Version 0.3.2121

**relevant settings**: 

**repository link (if public, optional)**: 

**code snippet to reproduce**: 
```rust
use aes::{
    cipher::{generic_array::GenericArray, KeyInit, KeyIvInit},
    Aes256,
};
use ofb::Ofb;

fn main() {
    let key = GenericArray::from_slice(&[0u8; 32]);
    let iv = GenericArray::from_slice(&[0u8; 16]);
    let ofb = Ofb::<Aes256>::new(key, iv);
    let ecb = Aes256::new(key);
}

```
Diagnostics show error at `Ofb::<Aes256>::new` : `expected 1 argument, found 2`. It believes `new` is from `KeyInit`, actually it should be from `KeyIvInit`.

**Comment by avalon1610** [comment]
I'm already using the latest version of `aes`.   
In fact, the problem lies with `Ofb::<Aes256>::new()`, not `Aes256::new()` and `cargo check` is fine.

![Image](https://github.com/user-attachments/assets/b5fe6411-cbb2-48b0-9ae4-e718b741b4ed)

**Comment by ChayimFriedman2** [comment]
@lnicola I can reproduce. I'm sure you're checking `Ofb::new()` and not `Aes256::new()`?

**Comment by lnicola** [comment]
It does happen now, see the underlined `iv` (compared to my previous screenshot).

![Image](https://github.com/user-attachments/assets/2685b3c0-93d5-42df-b7d5-e016d0e2b6a6)

🤷

**Comment by RivenSkaye** [comment]
Ended up here through RustCrypto/hashes#529 and I'm getting a false negative type identification as well. The MRE is rather simple and seems to be 
```rs
use sha2::{Digest, Sha512};

pub fn main() {
    let password = "Sup3rS3kr1tH4x0r";
    // This flags E0599 "no such associated item" as if RA isn't checking all valid impls
    println!("{:x}", Sha512::digest(password.as_bytes()));
    // This does not
    println!("{:x}", <Sha512 as Digest>::digest(password.as_bytes()));
}
```

**Comment by lnicola** [comment]
Both of these _appear_ work at the moment:

<img width="1213" height="488" alt="Image" src="https://github.com/user-attachments/assets/28bc65d8-31b2-424a-ac8e-efe642c09c00" />

<img width="897" height="227" alt="Image" src="https://github.com/user-attachments/assets/fde016f1-a845-48bb-acbc-0deb90c9d012" />
